netaddr.go

```diff
@@ -846,3 +894,374 @@ func (p *IPPrefix) UnmarshalText(text []byte) error {
 func (p IPPrefix) String() string {
 	return fmt.Sprintf("%s/%d", p.IP, p.Bits)
 }
+
+// LastIP returns the last IP in the prefix.
+func (p IPPrefix) LastIP() IP {
+	if p.IP.IsZero() {
+		return IP{}
+	}
+	a16 := p.IP.As16()
+	var off uint8
+	var bits uint8 = 128
+	if p.IP.Is4() {
+		off = 12
+		bits = 32
+	}
+	for b := p.Bits; b < bits; b++ {
+		byteNum, bitInByte := b/8, 7-(b%8)
+		a16[off+byteNum] |= 1 << uint(bitInByte)
+	}
+	if p.IP.Is4() {
+		return IPFrom16(a16)
+	} else {
+		return IPv6Raw(a16) // doesn't unmap
+	}
+}
+
+// IPRange represents an inclusive range of IP addresses
+// from the same address family.
+type IPRange struct {
+	// From is the starting IP address in the range.
+	// It must have the same address family (IPv4 vs IPv6) as To
+	// and be less than or equal to To.
+	// From is an inclusive bound; it is included in the set.
+	From IP
+
+	// To is the ending IP address in the range.
+	// It must have the same address family (IPv4 vs IPv6) as From
+	// and be greater than or equal to From.
+	// To is an inclusive bound; it is included in the set.
+	To IP
+}
+
+// Valid reports whether r.From and r.To are both non-zero and obey
+// the documented requirements: address families match, and From is
+// less than or equal to To.
+func (r IPRange) Valid() bool {
+	return !r.From.IsZero() && !r.To.IsZero() &&
+		r.From.Is4() == r.To.Is4() &&
+		!r.To.Less(r.From)
+}
+
+// ip16 represents a mutable IP address, either IPv4 (in IPv6-mapped
+// form) or IPv6.
+type ip16 [16]byte
+
+// bitSet reports whether the given bit in the address is set.
+// (bit 0 is the most significant bit in ip[0]; bit 127 is last)
+func (ip ip16) bitSet(bit uint8) bool {
+	i, s := bit/8, 7-(bit%8)
+	return ip[i]&(1<<s) != 0
+}
+
+func (ip *ip16) set(bit uint8) {
+	i, s := bit/8, 7-(bit%8)
+	ip[i] |= 1 << s
+}
+
+func (ip *ip16) clear(bit uint8) {
+	i, s := bit/8, 7-(bit%8)
+	ip[i] &^= 1 << s
+}
+
+// lastWithBitZero returns a copy of ip with the given bit
+// cleared and the following all set.
+func (ip ip16) lastWithBitZero(bit uint8) ip16 {
+	ip.clear(bit)
+	for ; bit < 128; bit++ {
```

Comment:
The potential optimization here made me go searching for a Go bit vector package, but amazingly, there doesn't seem to be a good one? Hard to believe.

Reply:
Using encoding/binary and then math/bits should generate pretty good code.

Reply:
Can optimize later.